Expand the sections regarding Jitter, Stutter and Latency

darksylinc opened the pull request:
This guide has basically two parts:

1. The result of thorough examination of all the problems found that were not caused by Godot (but were attributed to it). This adds an in-depth guide on how to locate and fix such issues.
2. A guide to the new code introduced by PR https://github.com/godotengine/godot/pull/106221 which still needs to be merged.

This is my first PR to godot-docs so please point out if I made a mistake and I apologize any mistake that could've been solved with a RTFM (I did read it, but I probably missed something).

Jordyfel commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+This table is read like this:
+ 
+ - If Waitable Swapchains are available, at "low" Godot will target 1 frame of latency. Otherwise, at "low" it will use AUTO as fallback.
+ - If Waitable Swapchains are available, at "low" Godot will use the value in frame_queue_size  as target for frames of latency. Otherwise it uses PARALLEL mode (which is the same as having no pacing method).
```
```suggestion
 - If Waitable Swapchains are available, at "medium" Godot will use the value in frame_queue_size  as target for frames of latency. Otherwise it uses PARALLEL mode (which is the same as having no pacing method).
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
 iOS devices are generally stutter-free, but older devices running newer versions
 of the operating system may exhibit problems. This is generally unavoidable.
 
+Troubleshooting Guide
+----------------------
+
+After an extensive evaluation of reported tickets and thorough testing of Godot on different HW,
+we found that Godot is blamed for many stutter, jitter or input lag that is not caused by Godot,
+but rather by 3rd Party Software or malfunctioning Hardware.
+
+This guide will help you find the root cause of these issues.
+
+**Please read it carefully and exhaust all options before reporting a ticket on Godot.**
```
```suggestion
Troubleshooting guide
---------------------

After an extensive evaluation of reported tickets and thorough testing of Godot on different hardware,
we found that Godot is blamed for many stutter, jitter or input lag that is not caused by Godot,
but rather by third-party software or malfunctioning hardware.

This guide will help you find the root cause of these issues.

**Please read it carefully and exhaust all options before reporting a ticket on Godot.**
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Bad HDMI / DisplayPort cable
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A broken cable may appear to function properly, but cause signal synchronization
+issues that go apparently unnoticed. However these issues can manifest when V-Sync
+is on and apps (including, but not limited, to Godot) running in exclusive fullscreen
+constantly have FPS jitter or slowdowns even in the most basic of scenes.
+
+If you see simple demo apps struggle to reach 60 fps (e.g. it reaches 60 fps then
+every 2 seconds slows down to 55 or 40 FPS and then goes back up) you may have a bad cable.
```
```suggestion
A broken cable may appear to function properly, but cause signal synchronization
issues that go apparently unnoticed. However, these issues can manifest when V-Sync
is enabled and apps (including but not limited to Godot) running in exclusive fullscreen
constantly have FPS jitter or slowdowns, even in the most basic of scenes.

If you see simple demo apps struggle to reach 60 FPS (e.g. it reaches 60 FPS, then
every 2 seconds slows down to 55 or 40 FPS and then goes back up), you may have a bad cable.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+**Workaround:** Disable V-Sync and use the FPS limiter by launching with: ``--disable-vsync --max-fps <fps>``
+
+Broken Monitor Firmware
+~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Broken monitor firmware
~~~~~~~~~~~~~~~~~~~~~~~
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Similar to a bad HDMI / DP cable, a poor interaction between your monitor firmware and
+your GPU can cause jitter or periodic slowdown.
+
+Unfortunately the simple answer is to replace the monitor. You might have luck contacting
+your Monitor manufacturer (e.g. LG, Samsung, Viewsonic, etc) and/or your GPU vendor
+(e.g. NVIDIA, AMD, Intel) and ask them for a solution or workaround.
```
```suggestion
Some monitors (typically the more recent/high-end ones) allow you to update their firmware. In this case, download the latest firmware and apply it.

For many other monitors, this is not an option. Unfortunately, in this case,
the simple answer is to replace the monitor. You might have luck contacting
your monitor manufacturer (e.g. LG, Samsung, ViewSonic, etc) and/or your GPU vendor
(e.g. NVIDIA, AMD, Intel) and ask them for a solution or workaround.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+your Monitor manufacturer (e.g. LG, Samsung, Viewsonic, etc) and/or your GPU vendor
+(e.g. NVIDIA, AMD, Intel) and ask them for a solution or workaround.
+
+**Solution:** Replace the monitor, or ask the manufacturer for a firmware/driver update.
```
```suggestion
**Solution:** Update firmware if possible, replace the monitor, or ask the manufacturer for a firmware/driver update.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+**Workaround:** Disable V-Sync and use the FPS limiter by launching with: ``--disable-vsync --max-fps <fps>``
+
+Multiple Monitors
+~~~~~~~~~~~~~~~~~
```
```suggestion
Multiple monitors
~~~~~~~~~~~~~~~~~
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+If you've got multiple monitors, ensure they are all in the same resolution, frequency (Hz),
+and bit depth (e.g. 16-bit vs 24-bit, HDR vs SDR).
+
+If they're not, change their settings until they match.
+
+If problems persist, try disabling all but one of the monitors.
```
```suggestion
If they're not, change their settings until they match. If problems persist, try disabling all but one of the monitors.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Overheating and throttling
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Jitter or stutter may be caused by an overheating CPU and/or GPU that throttles itself down.
+You can use temperature monitor software to see if this is the case.
```
Mention some examples of temperature monitors:

```suggestion
Jitter or stutter may be caused by an overheating CPU and/or GPU that throttles itself down.
You can use temperature monitor software such as
`Libre Hardware Monitor <https://github.com/LibreHardwareMonitor/LibreHardwareMonitor>`__ on Windows or
`lm-sensors <https://wiki.archlinux.org/title/Lm_sensors>`__ on Linux
to see if this is the case.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+**Solution:** Fix the cooling problem.
+
+Inconsistent Power Saving
+~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
Inconsistent power saving
~~~~~~~~~~~~~~~~~~~~~~~~~
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Inconsistent Power Saving
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In some cases the OS or driver keeps switching the CPU and/or GPU frequencies up and
+down; causing an uneven experience.
```
```suggestion
In some cases, the OS or driver keeps switching the CPU and/or GPU frequencies up and
down. This causes an uneven experience.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+    sudo cpupower frequency-set -g performance
+
+    # AMD
+    echo high | sudo tee /sys/class/drm/card1/device/power_dpm_force_performance_level
```
For future reference, this should be less of a necessity on a Linux distribution with a recent Mesa stack since https://gitlab.freedesktop.org/drm/amd/-/issues/1500 has been resolved. However, it can still improve performance (and latency) at times, so it's worth mentioning.

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+If Godot is not in one of the "bad" ones, there may be an overlay that is causing problems
+(e.g. AMD Overlay, NVIDIA's overlay, Microsoft Game Bar, "Activate Windows" watermark, etc).
+
+Godot can only enter the good ones if it's fully covering a single monitor with nothing on top.
+Some newer HW may be able to enter "Hardware Composed: Independent Flip" even if not fullscreen though.
```
```suggestion
Godot can only enter the good ones if it's fully covering a single monitor with nothing on top.
Some newer hardware may be able to enter "Hardware Composed: Independent Flip" even if not fullscreen though, thanks to support for multiple plane overlays (MPO).
This behavior can sometimes be unreliable (e.g. it may vary depending
on monitor count or whether HDR is enabled), so it's still better to
use fullscreen whenever possible.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+.. warning::
+
+    **Set PresentMon to windowed mode**. Otherwise PresentMon's overlay `ironically prevents apps <https://github.com/GameTechDev/PresentMon/issues/367>`__
+    from entering the "good" modes as it is displayed on top of Godot.
+    You will have to use the recording option and later see the CSV capture to see what mode Godot was in.
```
```suggestion
    **Set PresentMon to windowed mode**. Otherwise, PresentMon's overlay
    `ironically prevents apps <https://github.com/GameTechDev/PresentMon/issues/367>`__
    from entering the "good" modes as it is displayed on top of Godot.
    You will have to use the recording option and later see the CSV capture
    to see what mode Godot was in.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+.. warning::
+
+    Resources online assert that "Hardware Composed: Independent Flip" is the superior method of all.
+    However our measurements do not conclusively align with such claims. Ultimately the best experience
+    is done when all resources are entirely dedicated to one process. "Hardware Composed: Independent Flip"
+    suggests that resources are being diverted to display something else too, even if that something is
+    just giving CPU time to the DWM process.
```
```suggestion
    Resources online assert that "Hardware Composed: Independent Flip" is the superior method of all.
    However, our measurements do not conclusively align with such claims. Ultimately, the best experience
    is done when all resources are entirely dedicated to one process. "Hardware Composed: Independent Flip"
    suggests that resources are being diverted to display something else too, even if that something is
    just giving CPU time to the DWM process.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Windows: NVIDIA Presentation Method
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tip::
+
+    This section only applies to Vulkan and OpenGL. It does not affect D3D12.
```
```suggestion
.. note::

    This section only applies to Vulkan and OpenGL. It does not affect D3D12.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Linux: NVIDIA
+~~~~~~~~~~~~~
+
+On Linux machines using proprietary NVIDIA drivers, every time the process **nvidia-smi** is launched it
+causes a stutter. Unfortunately, many 3rd party apps periodically launch **nvidia-smi** to get GPU temperature
+and frequency and display it on an overlay or similar. For example running **nvidia-smi** once
+per second completely ruins the gaming experience.
```
```suggestion
On Linux machines using proprietary NVIDIA drivers, every time the process **nvidia-smi** is launched, it
causes a stutter. Unfortunately, many third-party apps periodically launch **nvidia-smi** to get GPU temperature
and frequency and display it on an overlay or similar. For example, running **nvidia-smi** once
per second completely ruins the gaming experience.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+and frequency and display it on an overlay or similar. For example running **nvidia-smi** once
+per second completely ruins the gaming experience.
+
+If an app can tell you GPU frequency, fan speed or temperature, then it is a suspect.
+
+Disabling or removing such apps will fix the problem.
```
```suggestion
If an app can tell you GPU frequency, fan speed or temperature, then it is a suspect.
Disabling or removing such apps will fix the problem.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+The following is a non-exhaustive list of known apps to cause stutter on NVIDIA Linux systems:
+
+ - `CPU-X <https://github.com/TheTumultuousUnicornOfDarkness/CPU-X>`__.
+ - Linux Mint's Mate panel application "CPU Frequency Scaling Monitor" applet.
+
+
+Linux: xfce
+~~~~~~~~~~~
```
```suggestion
 - `CPU-X <https://github.com/TheTumultuousUnicornOfDarkness/CPU-X>`__.
 - Linux Mint's Mate panel application "CPU Frequency Scaling Monitor" applet.

Linux: Xfce
~~~~~~~~~~~
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+Linux: xfce
+~~~~~~~~~~~
+
+xfce's compositor is not good for low latency games and can cause jitter. Disable it.
```
```suggestion
Xfce's compositor is not good for low-latency games and can cause jitter. Disable it.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,224 @@ iOS
+
+.. image:: img/xfce-disable-compositor.webp
+
+If you still want to use a Compositor on xfce, prefer using a better one like picom:
```
```suggestion
If you still want to use a compositor on Xfce, prefer using a better one like picom:
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
 On Linux, disabling compositing in window managers that allow it (such as KWin
 or Xfwm) can reduce input lag significantly.
 
+Latency Reduction
+~~~~~~~~~~~~~~~~~
```
```suggestion
Latency reduction
~~~~~~~~~~~~~~~~~
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+Latency Reduction
+~~~~~~~~~~~~~~~~~
+
+Starting with Godot 4.5, new pacing methods were introduced to reduce latency (and alleviate Jitter):
```
```suggestion
Starting with Godot 4.5, new pacing methods were introduced to reduce latency (and alleviate jitter):
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+  * ``PARALLEL``: The CPU and GPU try to run in parallel. This is the usual method of 2D and 3D rendering.
+  * ``AUTO`` automatically selects between SEQUENTIAL or PARALLEL based on current CPU and GPU performance. AUTO will only select SEQUENTIAL if the system is fast enough.
+
+Waitable Swapchains are by far the superior method, but if that's not available AUTO/SEQUENTIAL will be used instead.
```
```suggestion
Waitable Swapchains are by far the superior method, but if that's not available, ``AUTO``/``SEQUENTIAL`` will be used instead.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+
+.. warning::
+
+    In all cases (*including* Waitable Swapchain method), **lowering latency implies sacrificing framerate**.
+    The question is, how much FPS (frames per second) are you willing to sacrifice for latency. The relationship is not linear,
+    which means the lower the latency you want to achieve the greater the FPS sacrifice, which can get ridiculously high.
+    This is not a bug, it is just the nature of how it works.
```
```suggestion
    In all cases (*including* Waitable Swapchain method), **lowering latency implies sacrificing framerate**.
    The question is, how much FPS (frames per second) are you willing to sacrifice for latency. The relationship is not linear.
    This means the lower the latency you want to achieve, the greater the FPS sacrifice, which can become very significant.
    This is not a bug, it is just the nature of how it works.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+    Lowering latency makes Godot more susceptible to **microstutter**.
+    Thus if you're experiencing microstutter, use these options to *increase* latency instead.
+
+.. warning::
+    ``low_extreme`` might get you better latency (or even be a placebo) but always at a very large FPS cost.
+    Furthermore it may reintroduce microstutter. Caution is advised when seeking to lower latency too much.
+
+.. warning::
+    Just because ``low_extreme`` works great in your machine doesn't mean it will work fine
+    in other machines. Don't deploy to end users with this setting as a default.
```
Collapse the two warnings together:

```suggestion
.. warning::

    ``low_extreme`` might get you better latency (or even be a placebo) but always at a very large FPS cost.
    Furthermore, it may reintroduce microstutter. Caution is advised when seeking to lower latency too much.

    Just because ``low_extreme`` works great in your machine doesn't mean it will work fine
    in other machines. Don't deploy to end users with this setting as a default.
```

In general, we should avoid having multiple adominition blocks next to each other. Having 2 is sometimes acceptable, but we should definitely never have 3 or more.

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+   :widths: auto
+
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+   | Setting         | Target: Frames of latency (if Waitable Swapchains are available)| Waitable Swapchains Unavailable |
```
```suggestion
   | Setting         | Target: Frames of latency (if Waitable Swapchains are available)| Waitable swapchains unavailable |
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+   | Setting         | Target: Frames of latency (if Waitable Swapchains are available)| Waitable Swapchains Unavailable |
+   +=================+=================================================================+=================================+
+   | low_extreme     | 0                                                               | SEQUENTIAL                      |
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+   | low             | 1                                                               | AUTO                            |
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+   | medium          | rendering/rendering_device/vsync/frame_queue_size               | PARALLEL (Disabled)             |
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+   | high_throughput | Disabled                                                        | PARALLEL (Disabled)             |
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
```
```suggestion
   +=================+=================================================================+=================================+
   | low_extreme     | 0                                                               | ``SEQUENTIAL``                      |
   +-----------------+-----------------------------------------------------------------+---------------------------------+
   | low             | 1                                                               | ``AUTO``                            |
   +-----------------+-----------------------------------------------------------------+---------------------------------+
   | medium          | rendering/rendering_device/vsync/frame_queue_size               | ``PARALLEL`` (Disabled)             |
   +-----------------+-----------------------------------------------------------------+---------------------------------+
   | high_throughput | Disabled                                                        | ``PARALLEL`` (Disabled)             |
   +-----------------+-----------------------------------------------------------------+---------------------------------+
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+   | high_throughput | Disabled                                                        | PARALLEL (Disabled)             |
+   +-----------------+-----------------------------------------------------------------+---------------------------------+
+
+This table is read like this:
+ 
+ - If Waitable Swapchains are available, at "low" Godot will target 1 frame of latency. Otherwise, at "low" it will use AUTO as fallback.
+ - If Waitable Swapchains are available, at "medium" Godot will use the value in frame_queue_size  as target for frames of latency. Otherwise it uses PARALLEL mode (which is the same as having no pacing method).
+
+
+.. tip::
+
+    Godot's behavior on version 4.4 and earlier corresponds to ``high_throughput``.
```
```suggestion
This table is read like this:
 
 - If Waitable Swapchains are available, at "low", Godot will target 1 frame of latency. Otherwise, at "low" it will use AUTO as fallback.
 - If Waitable Swapchains are available, at "medium" Godot will use the value in ``frame_queue_size`` as target for frames of latency. Otherwise, it uses ``PARALLEL`` mode (which is the same as having no pacing method).

.. tip::

    Godot's behavior on version 4.4 and earlier corresponds to ``high_throughput``.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+But let's say we want to try another method. For some reason, we suspect waitable swapchains are malfunctioning.
+We can try masking this feature out:
+
+- ``--pacing-mode-mask 0x01`` will force to only use SEQUENTIAL_SYNC (if available).
+- ``--pacing-mode-mask 0x02`` will force to only use WAITABLE_SWAPCHAIN (if available).
```
```suggestion
- ``--pacing-mode-mask 0x01`` will force to only use ``SEQUENTIAL_SYNC`` (if available).
- ``--pacing-mode-mask 0x02`` will force to only use ``WAITABLE_SWAPCHAIN`` (if available).
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+   | ANDROID_SWAPPY     | 0x04 |
+   +--------------------+------+
+
+Thus by launching with ``--pacing-mode-mask 0x01`` we can experience how it runs by using ``SEQUENTIAL_SYNC``.
+Now that we are in sequential sync pacing mode; again we can try ``--latency-mode`` makes any difference.
+You can try ``low_extreme`` (SEQUENTIAL), ``low`` (AUTO) and ``medium`` (PARALLEL) to see if they make a difference.
```
```suggestion
Thus, by launching with ``--pacing-mode-mask 0x01`` we can experience how it runs by using ``SEQUENTIAL_SYNC``.
Now that we are in sequential sync pacing mode; again we can try ``--latency-mode`` makes any difference.
You can try ``low_extreme`` (``SEQUENTIAL``), ``low`` (``AUTO``) and ``medium`` (``PARALLEL``) to see if they make a difference.
```

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -241,6 +459,127 @@ If your mouse offers multiple :abbr:`DPI (Dots Per Inch)` settings, consider als
+Now that we are in sequential sync pacing mode; again we can try ``--latency-mode`` makes any difference.
+You can try ``low_extreme`` (SEQUENTIAL), ``low`` (AUTO) and ``medium`` (PARALLEL) to see if they make a difference.
+
+**When to report issues:**
+
+ - If WAITABLE_SWAPCHAIN at ``low`` or ``medium`` settings are causing problems. This could indicate an issue with the rendering API, the OS, or HW.
+ - If SEQUENTIAL_SYNC at ``low`` is causing problems but works fine at ``low_extreme`` or ``medium``. This would indicate a problem in Godot's AUTO algorithm.
+ - If ANDROID_SWAPPY is causing problems.
+
+**What NOT to report:**
+ - ``low`` or ``low_extreme`` reduces the FPS.
+ - ``AUTO`` does not improve latency.
+ - ``SEQUENTIAL_SYNC`` does not improve latency.
```
```suggestion
**When to report issues:**

 - If ``WAITABLE_SWAPCHAIN`` at ``low`` or ``medium`` settings are causing problems. This could indicate an issue with the rendering API, the OS, or HW.
 - If ``SEQUENTIAL_SYNC`` at ``low`` is causing problems but works fine at ``low_extreme`` or ``medium``. This would indicate a problem in Godot's AUTO algorithm.
 - If ``ANDROID_SWAPPY`` is causing problems.

**What NOT to report:**
 - ``low`` or ``low_extreme`` reduces the FPS.
 - ``AUTO`` does not improve latency.
 - ``SEQUENTIAL_SYNC`` does not improve latency.
```

Calinou reviewed (commented):
Looks great! I've made stylistic suggestions but the contents are pretty solid.

We should see if we can remake the Windows screenshots with fully English text later on, but it's not critical for getting this merged as the locations for the settings are obvious enough.

darksylinc commented:
Thanks @Calinou for the editorial. I've applied your fixes.

I've also added 2 more changes:

1. I added a note that nvidia-smi.exe exists on Windows too (but it is rarely a problem on Windows). Look for "nvidia-smi.exe" in the document.
2. Added "Third-party software" section.

> We should see if we can remake the Windows screenshots with fully English text later on, but it's not critical for getting this merged as the locations for the settings are obvious enough.

Agreed. I tried to get them in full English but some text stayed in Spanish even after installing the English language pack and switching to it.

Calinou commented on tutorials/rendering/jitter_stutter.rst:
```diff
@@ -147,6 +147,261 @@ iOS
+- `Game Bar <https://www.microsoft.com/windows/tips/game-bar>`__ as previously covered, Game Bar may prevent Godot from entering "Hardware Independent Flip". Disable it to see if that fixes problems.
+- `Discord Game Overlay <https://support.discord.com/hc/en-us/articles/217659737-Game-Overlay-101#h_01JPT35B3CA450SKEF4N9DX70R>`__. Same deal as with the other overlays.
+- WeChat / WeGame is known to cause problems with Godot (`#83307 <https://github.com/godotengine/godot/issues/83307>`__, `#81652 <https://github.com/godotengine/godot/issues/81652>`__). Either uninstall it, remove the implicit Vulkan Layer, or use the `VK_LOADER_LAYERS_DISABLE environment variable <https://github.com/godotengine/godot/issues/81652#issuecomment-2510962586>`__ to disable it.
+- RGB Led control software. A lot of software to control RGB lights (whether it's the keyboard, mouse, RAM sticks, fans, GPU, etc) are poorly written and known to cause stutter and jitter problems. Uninstall them. Prefer purchasing devices that are compatible with `Windows 11' Dynamic Lighting <https://support.microsoft.com/en-us/windows/control-dynamic-lighting-devices-in-windows-8e8f22e3-e820-4 ...
```
```suggestion
- RGB LED control software. A lot of software to control RGB lights (whether it's the keyboard, mouse, RAM sticks, fans, GPU, etc) are poorly written and known to cause stutter and jitter problems. Uninstall them. Prefer purchasing devices that are compatible with `Windows 11' Dynamic Lighting <https://support.microsoft.com/en-us/windows/control-dynamic-lighting-devices-in-windows-8e8f22e3-e820-476c-8f9d-9ffc7b6ffcd2>`__ control.
```

Calinou reviewed (approved):
Should be good to merge after applying suggestions and rebasing.
